Document Xapi's auto-generated modules

contificate opened the pull request:
Provides a short article that tries to elucidate aspects concerning API, Server, Client, and Db_actions.

![image](https://github.com/user-attachments/assets/08f4c7ef-53c7-48d8-b75d-26a0707ca0ab)

contificate commented:
Opening as a draft as I would appreciate feedback on the correctness and style of my presentation. These details are quite important, I think, and I believe documenting the concrete internals (even if they've been cemented for many years) is important for new developers.

---

I also intend to write about some more things (e.g. `Rbac_static` is also auto-generated and has important semantics w.r.t acting as a database) at some point.

edwintorok commented on doc/content/xapi/internals/generated.md:
```diff
@@ -0,0 +1,345 @@
+
+## Interface Description
+
+All of XenAPI's data model is described within the `ocaml/idl` subdirectory.
```
Not to be confused with `xapi-idl`, which describes the internal APIs in a different form

contificate commented on doc/content/xapi/internals/generated.md:
```diff
@@ -0,0 +1,345 @@
+important for `Server` - described later - which receives that format as
+input).
+
+## Database Actions (`db_actions.ml`)
```
- It would be worth adding some commentary about the handling of `(_ ref) set` fields which, as I recall, are handled in `Db_actions` (there's a partition when doing `get_record`).

- Also add some commentary about `*_init` modules in `Db_actions` which register themselves with `Eventgen` to allow `Xapi_event` (through a few hoops) to collect snapshots for event handling.

last-genius replied:
I think, given that this is a developer-facing resource, it would be worthwhile to include these TODOs in the text itself if you don't manage to come up with something reasonable about this, as they could still serve as useful pointers.

contificate replied:
Hopefully managed to address this commentary with my latest push. If I think there's any other detail that I haven't explained, but is at least worth mentioning, I'll leave a stray TODO-esque comment.

contificate commented on doc/content/xapi/internals/generated.md:
```diff
@@ -0,0 +1,345 @@
+transport that may be asynchronous (for example, within the context of a
+program using `Lwt` or `Async` for its networking).
+
+There is also a sub-functor `AsyncF` (within `ClientF`) that is parameterised
```
This is hard to follow, the nesting of these modules could be explained better by a code snippet that replaces a lot of detail with comments.

contificate replied:
Added a snippet that tries to show the structure more clearly.

contificate commented:
Added a small section on `Rbac_static`, explained some of the role of `Eventgen` w.r.t `Db_actions`, and introduced a snippet that illustrates the shape of the nested functors within `Client`.

contificate commented:
I think the current contents are a suitable start for the document (thanks for the approvals). We can always come back to it if we find something needs further clarification, or a new section.
